umi_tools/umi_methods.py

```diff
@@ -1104,6 +1104,80 @@ def get_read_position(read, soft_clip_threshold):
     return start, pos, is_spliced
 
 
+class get_readpairs:
+    # Used in place of a real read as the mate of reads that aren't properly paired
+    NoMate = object()
+
+    def incomplete_pair_to_singleton(self, queue_entry):
+        for i in [i for i, v in enumerate(queue_entry) if v is get_readpairs.NoMate]:
+            queue_entry[i] = None
+
+    def incomplete_pairs_as_singletons(self):
+        self.incomplete_queue_entries.clear()
+        while self.queue:
+            queue_entry = self.queue.popleft()
+            if get_readpairs.NoMate in queue_entry:
+                # Incomplete entry. Turn into singleton entry
+                read = next(r for r in queue_entry if r is not get_readpairs.NoMate)
+                U.warn("inconsistent BAM file: only one of the two mates of proper pair %s found on chromosome %s, treating as unpaired" %
+                       (read.qname, read.reference_name))
+                incomplete_pair_to_singleton(queue_entry)
+            yield queue_entry
+
+    def __call__(self, inreads):
+        self.queue = collections.deque()
+        self.incomplete_queue_entries = dict()
+        self.current_chr = None
+
+        for read in inreads:
+            read_i = int(read.is_read2)
+            read_chr = read.reference_name
+
+            # Output leftover incomplete read pairs if the chrosomome changes
+            if self.current_chr is not None and self.current_chr != read_chr:
+                for queue_entry in self.incomplete_pairs_as_singletons():
+                    yield queue_entry
+            self.current_chr = read_chr
+
+            # Queue read, either as a singleton, or as part of a read pair
+            if read.is_unmapped or read.mate_is_unmapped or not read.is_proper_pair:
+                # Read is not part of a proper pair. Set mate to None
+                queue_entry = [None, None]
+                queue_entry[read_i] = read
+                self.queue.append(queue_entry)
+            elif read.qname in self.incomplete_queue_entries:
+                # Read is part of an already queued (incomplete) read pair. Complete it.
+                queue_entry = self.incomplete_queue_entries.pop(read.qname)
+                if queue_entry[read_i] is not get_readpairs.NoMate:
+                    U.warn("inconsistent BAM file: both mates %s flagged to be read%d" %
```

Comment:
Is is going to be a problem with multi mappers. Many people allow multimappers, and they are unavoidable when mapping to the transcriptome.

Reply:
True, but again I don't really see how multi-mapping and paired reads should work together at all. The fundamental problem seems to be that there are n*n possible pairing if both read1 and read2 are mapped to n locations. And the BAM format at best may tell us which the "primary" pair is, but not how the secondary mapping should be paired. Maybe I'm missing the point, though -- I haven't really worked with multi-mapped data so far...

Reply:
On second thought, I could do the same that `TwoPassPairWriter` does, and use reference name and mapping position as additional keys to identify mate pairs. Would that alleviate these concerns?